### ISSUE TITLE
[inputs.mqtt] some data missing in database

### ISSUE THREAD

**Issue by vielleicht** [influxdata/telegraf#14749]
### Relevant telegraf.conf

```toml
[[inputs.mqtt_consumer]]
  name_override = "mqtt_consumer_dict"
  servers = ["tcp://172.19.0.1:1883"]

  qos = 0
  connection_timeout = "30s"

  topics = [
    #...
    "shellyplusplugs-dose3/status/switch:0"
    #...
  ]
  persistent_session = true
  client_id = "telegraf-server2"
  data_format = "json"

[[outputs.influxdb_v2]]
    urls = ["http://172.19.0.4:8086"]
    token = "..."
    organization = "..."
    bucket = "70e8bf123f145d99"
```

### Logs from Telegraf

```text
telegraf  | 2024-02-11T14:05:39Z I! Loading config: /etc/telegraf/telegraf.conf
telegraf  | 2024-02-11T14:05:39Z W! DeprecationWarning: Option "perdevice" of plugin "inputs.docker" deprecated since version 1.18.0 and will be removed in 2.0.0: use 'perdevice_include' instead
telegraf  | 2024-02-11T14:05:39Z I! Starting Telegraf 1.29.2 brought to you by InfluxData the makers of InfluxDB
telegraf  | 2024-02-11T14:05:39Z I! Available plugins: 241 inputs, 9 aggregators, 30 processors, 24 parsers, 60 outputs, 6 secret-stores
telegraf  | 2024-02-11T14:05:39Z I! Loaded inputs: cpu disk diskio docker kernel mem mqtt_consumer (3x) processes swap system systemd_units
telegraf  | 2024-02-11T14:05:39Z I! Loaded aggregators:
telegraf  | 2024-02-11T14:05:39Z I! Loaded processors:
telegraf  | 2024-02-11T14:05:39Z I! Loaded secretstores:
telegraf  | 2024-02-11T14:05:39Z I! Loaded outputs: influxdb_v2
telegraf  | 2024-02-11T14:05:39Z I! Tags enabled: host=kerschbaam
telegraf  | 2024-02-11T14:05:39Z W! Deprecated inputs: 0 and 1 options
telegraf  | 2024-02-11T14:05:39Z I! [agent] Config: Interval:10s, Quiet:false, Hostname:"kerschbaam", Flush Interval:1m0s
telegraf  | 2024-02-11T14:05:39Z D! [agent] Initializing plugins
telegraf  | 2024-02-11T14:05:39Z D! [agent] Connecting outputs
telegraf  | 2024-02-11T14:05:39Z D! [agent] Attempting connection to [outputs.influxdb_v2]
telegraf  | 2024-02-11T14:05:39Z D! [agent] Successfully connected to outputs.influxdb_v2
telegraf  | 2024-02-11T14:05:39Z D! [agent] Starting service inputs
telegraf  | 2024-02-11T14:05:39Z I! [inputs.mqtt_consumer] Connected [tcp://172.19.0.1:1883]
telegraf  | 2024-02-11T14:05:39Z D! [inputs.mqtt_consumer] Session found [tcp://172.19.0.1:1883]
telegraf  | 2024-02-11T14:05:39Z I! [inputs.mqtt_consumer] Connected [tcp://172.19.0.1:1883]
telegraf  | 2024-02-11T14:05:39Z D! [inputs.mqtt_consumer] Session found [tcp://172.19.0.1:1883]
telegraf  | 2024-02-11T14:05:39Z I! [inputs.mqtt_consumer] Connected [tcp://172.19.0.1:1883]
telegraf  | 2024-02-11T14:05:39Z D! [inputs.mqtt_consumer] Session found [tcp://172.19.0.1:1883]
telegraf  | 2024-02-11T14:05:40Z E! [inputs.systemd_units] Error in plugin: exec: "systemctl": executable file not found in $PATH
telegraf  | 2024-02-11T14:05:40Z D! [inputs.system] Reading users: open /var/run/utmp: no such file or directory
telegraf  | 2024-02-11T14:05:40Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/sys/firmware/efi/efivars"): no such file or directory
telegraf  | 2024-02-11T14:05:40Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/run/docker/netns/0f9dfd4d4805"): no such file or directory
telegraf  | 2024-02-11T14:05:40Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/run/docker/netns/5aaf117ac444"): no such file or directory
telegraf  | 2024-02-11T14:05:40Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/run/docker/netns/971c78a6358e"): no such file or directory
telegraf  | 2024-02-11T14:05:40Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/run/docker/netns/57ec8245f222"): no such file or directory
telegraf  | 2024-02-11T14:05:40Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/run/docker/netns/1ba835b4d3a8"): no such file or directory
telegraf  | 2024-02-11T14:05:40Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/run/docker/netns/2fe7fbd63243"): no such file or directory
telegraf  | 2024-02-11T14:05:40Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/run/docker/netns/8cc348eb17d7"): no such file or directory
telegraf  | 2024-02-11T14:05:40Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/run/docker/netns/4ac8215eedb4"): no such file or directory
telegraf  | 2024-02-11T14:05:40Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/run/docker/netns/d4160e4c98c1"): no such file or directory
telegraf  | 2024-02-11T14:05:40Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/run/docker/netns/0fc16b414b9d"): no such file or directory
telegraf  | 2024-02-11T14:05:43Z D! [inputs.mqtt_consumer] No metrics were created from a message. Verify your parser settings. This message is only printed once.
telegraf  | 2024-02-11T14:05:43Z E! [inputs.mqtt_consumer] Error in plugin: strconv.ParseFloat: parsing "true": invalid syntax
telegraf  | 2024-02-11T14:05:43Z E! [inputs.mqtt_consumer] Error in plugin: strconv.ParseFloat: parsing "heat": invalid syntax
telegraf  | 2024-02-11T14:05:43Z E! [inputs.mqtt_consumer] Error in plugin: strconv.ParseFloat: parsing "heating": invalid syntax
telegraf  | 2024-02-11T14:05:46Z E! [inputs.mqtt_consumer] Error in plugin: strconv.ParseFloat: parsing "true": invalid syntax
telegraf  | 2024-02-11T14:05:49Z E! [inputs.mqtt_consumer] Error in plugin: strconv.ParseFloat: parsing "true": invalid syntax
telegraf  | 2024-02-11T14:05:50Z E! [inputs.systemd_units] Error in plugin: exec: "systemctl": executable file not found in $PATH
telegraf  | 2024-02-11T14:05:50Z D! [inputs.system] Reading users: open /var/run/utmp: no such file or directory
telegraf  | 2024-02-11T14:05:50Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/sys/firmware/efi/efivars"): no such file or directory
telegraf  | 2024-02-11T14:05:50Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/run/docker/netns/0f9dfd4d4805"): no such file or directory
telegraf  | 2024-02-11T14:05:50Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/run/docker/netns/5aaf117ac444"): no such file or directory
telegraf  | 2024-02-11T14:05:50Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/run/docker/netns/971c78a6358e"): no such file or directory
telegraf  | 2024-02-11T14:05:50Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/run/docker/netns/57ec8245f222"): no such file or directory
telegraf  | 2024-02-11T14:05:50Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/run/docker/netns/1ba835b4d3a8"): no such file or directory
telegraf  | 2024-02-11T14:05:50Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/run/docker/netns/2fe7fbd63243"): no such file or directory
telegraf  | 2024-02-11T14:05:50Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/run/docker/netns/8cc348eb17d7"): no such file or directory
telegraf  | 2024-02-11T14:05:50Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/run/docker/netns/4ac8215eedb4"): no such file or directory
telegraf  | 2024-02-11T14:05:50Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/run/docker/netns/d4160e4c98c1"): no such file or directory
telegraf  | 2024-02-11T14:05:50Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/run/docker/netns/0fc16b414b9d"): no such file or directory
telegraf  | 2024-02-11T14:05:52Z E! [inputs.mqtt_consumer] Error in plugin: strconv.ParseFloat: parsing "true": invalid syntax
telegraf  | 2024-02-11T14:05:55Z E! [inputs.mqtt_consumer] Error in plugin: strconv.ParseFloat: parsing "true": invalid syntax
telegraf  | 2024-02-11T14:05:58Z E! [inputs.mqtt_consumer] Error in plugin: strconv.ParseFloat: parsing "true": invalid syntax
telegraf  | 2024-02-11T14:06:00Z E! [inputs.systemd_units] Error in plugin: exec: "systemctl": executable file not found in $PATH
telegraf  | 2024-02-11T14:06:00Z D! [inputs.system] Reading users: open /var/run/utmp: no such file or directory
telegraf  | 2024-02-11T14:06:00Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/sys/firmware/efi/efivars"): no such file or directory
telegraf  | 2024-02-11T14:06:00Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/run/docker/netns/0f9dfd4d4805"): no such file or directory
telegraf  | 2024-02-11T14:06:00Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/run/docker/netns/5aaf117ac444"): no such file or directory
telegraf  | 2024-02-11T14:06:00Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/run/docker/netns/971c78a6358e"): no such file or directory
telegraf  | 2024-02-11T14:06:00Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/run/docker/netns/57ec8245f222"): no such file or directory
telegraf  | 2024-02-11T14:06:00Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/run/docker/netns/1ba835b4d3a8"): no such file or directory
telegraf  | 2024-02-11T14:06:00Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/run/docker/netns/2fe7fbd63243"): no such file or directory
telegraf  | 2024-02-11T14:06:00Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/run/docker/netns/8cc348eb17d7"): no such file or directory
telegraf  | 2024-02-11T14:06:00Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/run/docker/netns/4ac8215eedb4"): no such file or directory
telegraf  | 2024-02-11T14:06:00Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/run/docker/netns/d4160e4c98c1"): no such file or directory
telegraf  | 2024-02-11T14:06:00Z D! [inputs.disk] [SystemPS] => unable to get disk usage ("/run/docker/netns/0fc16b414b9d"): no such file or directory
telegraf  | 2024-02-11T14:06:01Z E! [inputs.mqtt_consumer] Error in plugin: strconv.ParseFloat: parsing "true": invalid syntax
telegraf  | 2024-02-11T14:06:02Z D! [outputs.influxdb_v2] Wrote batch of 1000 metrics in 53.662319ms
telegraf  | 2024-02-11T14:06:02Z D! [outputs.influxdb_v2] Buffer fullness: 0 / 10000 metrics
```

### System info

Starting Telegraf 1.29.2, Linux version 6.7.2-arch1-1 (linux@archlinux), Docker version 24.0.7

### Docker

```
services:
  influxdb:
    image: influxdb:2.7.5-alpine
    container_name: influxdb
    volumes:
      - ./influxdb2:/var/lib/influxdb2:rw
      - ./influxdb2:/etc/influxdb2
      - ./telegraf.conf:/etc/telegraf.conf:rw
    restart: on-failure:10
    ports:
      - 8086:8086
    environment:
      INFLUXDB_LOGGING_LEVEL: debug

  telegraf:
    image: telegraf:1.29.2-alpine
    container_name: telegraf
    volumes:
      - ./telegraf.conf:/etc/telegraf/telegraf.conf:ro
      - /var/run/docker.sock:/var/run/docker.sock:ro
      - /sys:/rootfs/sys:ro
      - /proc:/rootfs/proc:ro
      - /etc:/rootfs/etc:ro
      - /sys/firmware/efi/efivars:/sys/firmware/efi/efivars:ro
      - /dev/hugepages:/dev/hugepages:ro
      - /boot:/boot:ro
      - /data:/data:ro
    environment:
      HOST_PROC: /rootfs/proc
      HOST_SYS: /rootfs/sys
      HOST_ETC: /rootfs/etc
    hostname: kerschbaam
    depends_on:
      - influxdb
    # find docker group by: $(stat -c '%g' /var/run/docker.sock)
    user: telegraf:972
```

### Steps to reproduce

1. start docker containers (expects a brocker to be running)
2. publish a MQTT message on topic `shellyplusplugs-dose3/status/switch:0` and content:
```
{"id":0, "source":"button", "output":true, "apower":33.4, "voltage":238.3, "current":0.194, "aenergy":{"total":777.605,"by_minute":[94.757,638.341,624.474],"minute_ts":1707660729},"temperature":{"tC":40.8, "tF":105.4}}
```

### Expected behavior

The message should be parsed and inserted into InfluxDB.

### Actual behavior

Message seems not be inserted into influx output. Maybe `[inputs.mqtt_consumer] No metrics were created from a message. Verify your parser settings. This message is only printed once.` is a hint but I cannot debug, which plugin is producing this.

### Additional info

There are more plugins running and some are causing issues. BUT the configured plugin is inserting metrics from different topics, I checked for this. I also changed the `name_override` and checked if there is some error in the database preventing the metrics to be written. But, the new table was created and metrics have been inserted.

I suspect, there is some problem with the `inputs.mqtt_consumer` plugin using the `data_format = "json"` I'm using. I would love to give more info, if needed to debug.

**Comment by vielleicht** [comment]
For debugging purposes I added an `outputs.file` config to telegraf, but the expected data isn't visible there either. So I once more would tend to blame `inputs.mqtt`...

```
[[outputs.file]]
  files = ["/tmp/metrics.out"]
  data_format = "influx"
```

I also tried `data_format = "json"` and `csv` to exclude output serializer from the list possible bugs.

**Comment by vielleicht** [comment]
I now tried `client_trace = true`, but did't find the log output very helpful:
```
telegraf  | 2024-02-11T20:26:14Z D! [inputs.mqtt_consumer] [net]     startIncoming Received Message
telegraf  | 2024-02-11T20:26:14Z D! [inputs.mqtt_consumer] [net]     startIncomingComms: got msg on ibound
telegraf  | 2024-02-11T20:26:14Z D! [inputs.mqtt_consumer] [net]     startIncomingComms: received publish, msgId:0
telegraf  | 2024-02-11T20:26:14Z D! [inputs.mqtt_consumer] [net]     logic waiting for msg on ibound
telegraf  | 2024-02-11T20:26:14Z D! [inputs.mqtt_consumer] [net]     startIncoming Received Message
telegraf  | 2024-02-11T20:26:14Z D! [inputs.mqtt_consumer] [net]     startIncomingComms: got msg on ibound
telegraf  | 2024-02-11T20:26:14Z D! [inputs.mqtt_consumer] [net]     startIncomingComms: received publish, msgId:0
telegraf  | 2024-02-11T20:26:14Z D! [inputs.mqtt_consumer] [net]     logic waiting for msg on ibound
telegraf  | 2024-02-11T20:26:14Z D! [inputs.mqtt_consumer] [net]     startIncoming Received Message
telegraf  | 2024-02-11T20:26:14Z D! [inputs.mqtt_consumer] [net]     startIncomingComms: got msg on ibound
telegraf  | 2024-02-11T20:26:14Z D! [inputs.mqtt_consumer] [net]     startIncomingComms: received publish, msgId:0
telegraf  | 2024-02-11T20:26:14Z D! [inputs.mqtt_consumer] [net]     logic waiting for msg on ibound
telegraf  | 2024-02-11T20:26:14Z D! [inputs.mqtt_consumer] [net]     startIncoming Received Message
telegraf  | 2024-02-11T20:26:14Z D! [inputs.mqtt_consumer] [net]     startIncomingComms: got msg on ibound
telegraf  | 2024-02-11T20:26:14Z D! [inputs.mqtt_consumer] [net]     startIncomingComms: received publish, msgId:0
telegraf  | 2024-02-11T20:26:14Z D! [inputs.mqtt_consumer] [net]     logic waiting for msg on ibound
telegraf  | 2024-02-11T20:26:14Z D! [inputs.mqtt_consumer] [net]     startIncoming Received Message
telegraf  | 2024-02-11T20:26:14Z D! [inputs.mqtt_consumer] [net]     startIncomingComms: got msg on ibound
telegraf  | 2024-02-11T20:26:14Z D! [inputs.mqtt_consumer] [net]     startIncomingComms: received publish, msgId:0
```

and so on and so forth... I hoped for some more usefull output.

**Comment by vielleicht** [comment]
I found the problematic config, after debugging mosquitto. I stumbled across missing subscriptions from the telegraf client. I then found the `persistent_session = true` leading to reusage of the old subscriptions.

Would be good, if telegraf could detect a changing configuration and handle updating the subscriptions. I would love to use the feature of the broker saving messages during connection loss.
